metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery_audit.py

```diff
@@ -84,7 +84,19 @@ class BigqueryTableIdentifier:
 
     @staticmethod
     def get_table_and_shard(table_name: str) -> Tuple[str, Optional[str]]:
-        match = re.search(
+        """
+        Args:
+            table_name:
+                table name (in form <table-id> or <table-prefix>_<shard>`, optionally prefixed with <project-id>.<dataset-id>)
+                See https://cloud.google.com/bigquery/docs/reference/standard-sql/wildcard-table-reference
+                If table_name is fully qualified, i.e. prefixed with <project-id>.<dataset-id> then the special case of
+                dataset as a sharded table, i.e. table-id itself as shard can not be detected.
+        Returns:
+            (table_name_without_shard, shard):
+                In case of non-sharded tables, returns (<table-id>, None)
+                In case of sharded tables, returns (<table-prefix>, shard)
+        """
+        match = re.match(
```

Comment:
Note the change from re.search to re.match (match the end of string with _). This is done to differentiate

table_2023010920230109 (as non-sharded table, as ) from
table_20230109 (as sharded table).

Reply:
In that case, we should add `$` to the regex to mark the end of the string but keep it as re.search

That way this method will work for fully qualified table names too

Reply:
I am having trouble updating the single regex to work for both qualified table name as well as non-qualified table name and support all the cases, at the same time. If this change does not look okay, I can revert this (and keep only others) and we can fix this in follow up PRs.

Reply:
Ok that's fine then - we can do it in a follow up PR

The reason I'd like to do it is that some folks partition at the dataset level instead of the table level, so passing the full name into this method enables us to support them in the future

Reply:
It'd be nice to simplify this logic as much as possible, and thus try to pass as few types of table identifiers as possible. Most of the time we pass fully qualified names or refs, so ideally I think we try to stop passing just table name most of the time? But am fine with this as Harshal is for now